src/os/bluestore/BlueStore.h

```diff
@@ -1644,6 +1645,186 @@ class BlueStore : public ObjectStore,
+      uint64_t blob_offset,
+      uint64_t length) = 0;
+
+    int _verify_csum(
+      const ghobject_t& oid,
+      const bluestore_blob_t* blob,
+      uint64_t blob_xoffset,
+      const ceph::bufferlist& bl,
+      uint64_t logical_offset);
+
+  public:
+    typedef std::unique_ptr<RegionReader> Ref;
+    typedef std::function<int(uint64_t,		//< offset
+                              uint64_t,		//< length
+                              ceph::bufferlist*,//< buffer to fill
+                              size_t)> io_issuer_t;
+
+    RegionReader(BlueStore* const store, BlobRef bptr)
+      : store(store),
+        bptr(bptr) {
+    }
+    virtual ~RegionReader() {
+    }
+
+    virtual int issue_io(io_issuer_t) = 0;
+    virtual int postprocess(ready_regions_t&, bool buffered) = 0;
+
+    static Ref create_or_update(
+      Ref&& current,
+      BlueStore* store,
+      BlobRef blob,
+      uint64_t logical_offset,
+      uint64_t blob_offset,
+      uint64_t length);
+  };
+
+  typedef RegionReader::Ref RegionReaderRef;
+
+  class PlainRegionReader : public RegionReader {
+    // --------------------------------------------------------
+    // intermediate data structures used while reading non-compressed
+    // portions of an object
+    struct plain_region_t : public region_t {
+      ceph::bufferlist bl;
+
+      // used later in read process
+      uint64_t front = 0;
+      uint64_t r_off = 0;
+
+      using region_t::region_t;
+    };
+
+    boost::container::small_vector<plain_region_t, 1> regions2read;
+
+    void add_region_hint(
+      const uint64_t logical_offset,
+      const uint64_t blob_offset,
+      const uint64_t length) override {
+      regions2read.emplace_back(logical_offset, blob_offset, length);
+    }
+
+public:
+    using RegionReader::RegionReader;
+    int issue_io(io_issuer_t) override;
+    int postprocess(ready_regions_t&, bool buffered) override;
+  };
+
+  class CompressedRegionReader : public RegionReader {
+    boost::container::small_vector<region_t, 1> regions2read;
+    ceph::bufferlist compressed_bl;
+   
+    void add_region_hint(
+      const uint64_t logical_offset,
+      const uint64_t blob_offset,
+      const uint64_t length) override {
+      regions2read.emplace_back(logical_offset, blob_offset, length);
+    }
+
+    boost::optional<ceph::bufferlist> _decompress(
+      ceph::bufferlist& source);
+
+  public:
+    using RegionReader::RegionReader;
+    int issue_io(io_issuer_t) override;
+    int postprocess(ready_regions_t&, bool buffered) override;
+  };
+
+  typedef std::map<BlobRef,
+                   RegionReaderRef> blobs2read_t;
+
+  struct cache_response_t {
+    ready_regions_t ready_regions;
+    blobs2read_t blobs2read;
+    size_t num_all_regions;
+  };
+  
+  cache_response_t _consult_cache(
+    BlueStore::OnodeRef o,
+    const uint64_t offset,
+    const size_t length);
+
+  enum class HoleMode {
+    ZEROIZE,
+    INDEX
+  };
+
+  struct AioReadBatch : public AioContext {
```

Comment:
Hmm, we really might want to make `AioReadBatch` (and probably other derivatives of `AioContext` as well) `final` to let compiler inline and optimize the destructor call at the end of `AioReadBatch::aio_finish()`. At the moment we're burning a lot of CPU cycles resetting `VPTR` to fulfill [the language restriction for dynamic polymorphism in ctor/dtor](https://isocpp.org/wiki/faq/strange-inheritance#calling-virtuals-from-ctors):

```BlueStore::AioReadBatch::~AioReadBatch  /usr/local/bin/ceph-osd
 35,42 │930ca0:   lea    0xb677d9(%rip),%rax        # 1498480 <vtable for BlueStore::AioReadBatch>
 27,78 │930ca7:   push   %r12
       │930ca9:   mov    %rdi,%r12
       │930cac:   push   %rbp
  1,39 │930cad:   push   %rbx
  1,39 │930cae:   add    $0x160,%rdi
       │930cb5:   add    $0x10,%rax
       │930cb9:   mov    %rax,-0x160(%rdi)
```